### PR TITLE
chore: upgrade Node.js and pnpm

### PR DESCRIPTION
**Pull request by kotarella1110**
- Node.js Upgrade: upgrading our Node.js version to the latest LTS release.
- pnpm Upgrade: updating our pnpm version from v7 to v8.
- Codesandbox CI Build Issue Resolution: This PR also resolves the issue causing Codesandbox CI builds to fail, ensuring that our continuous integration pipeline runs smoothly and consistently.

These updates collectively aim to enhance the reliability, and maintainability of our project while also addressing the specific issue related to Codesandbox CI.

### CODE REVIEW

File: .codesandbox/ci.json [hunk 1]
```diff
@@ -1,8 +1,9 @@
 {
+  "installCommand": "csb:install",
   "sandboxes": [
     "react-hook-form-v7-js-template-quqov",
     "react-hook-form-v7-ts-template-4tn8i"
   ],
-  "node": "14",
+  "node": "18",
```
Comment (kotarella1110) [inline .codesandbox/ci.json:7]: Codesandbox CI does not support Node.js 20, so I am specifying version 18 here.

**Comment by kotarella1110** [comment]
The workflow that compares package sizes is failing because the composite action added in this PR does not exist in the base branch. Merging this PR will resolve the issue and allow the workflow to pass without any problems.
I have confirmed that checking out from this branch to create a PR https://github.com/react-hook-form/react-hook-form/pull/11164 and running the package size comparison workflow results in success.

File: cypress/e2e/setValueAsyncStrictMode.cy.ts [hunk 2]
```diff
@@ -2,7 +2,7 @@ describe('form setValueAsyncStrictMode', () => {
   it('should set async input value correctly', () => {
     cy.visit('http://localhost:3000/setValueAsyncStrictMode');
 
-    cy.wait(10);
+    cy.wait(100);
```
Comment (bluebill1049) [inline cypress/e2e/setValueAsyncStrictMode.cy.ts:5]: 100?

**Review by bluebill1049 — APPROVED** [review]

**Comment by bluebill1049** [comment]
nice one! looks like build step failed

**Comment by kotarella1110** [comment]
> nice one! looks like build step failed

@bluebill1049 

The package size comparison workflow is [failing](https://github.com/react-hook-form/react-hook-form/actions/runs/6785593975/job/18444263152?pr=11163) because Because the version of pnpm is different between the master branch and the PR branch. Merging this PR will resolve the issue and allow the workflow to pass without any problems.
I have confirmed that checking out from this branch to create a PR https://github.com/react-hook-form/react-hook-form/pull/11164 and running the package size comparison workflow results in success.

File: cypress/e2e/setValueAsyncStrictMode.cy.ts (hunk 2 above)
Comment (kotarella1110) [reply cypress/e2e/setValueAsyncStrictMode.cy.ts:5]: This test began to fail randomly and was solved by extending the delay time.

**Pull request merged**